src/main/cpp/src/mixed_sort_merge_join.cu

```diff
@@ -0,0 +1,825 @@
+    auto output_dest                = cudf::ast::detail::value_expression_result<bool, has_nulls>();
+    cudf::size_type const left_idx  = left_indices[i];
+    cudf::size_type const right_idx = right_indices[i];
+
+    evaluator.evaluate(output_dest, left_idx, right_idx, 0, thread_intermediate_storage);
+
+    keep_mask[i] = output_dest.is_valid() && output_dest.value();
+  }
+}
+
+/**
+ * @brief Filter join results by evaluating conditional expression
+ *
+ * This function takes references to the input index vectors (not rvalue references) and uses
+ * exec_policy_nosync for all thrust operations. The caller is responsible for synchronizing
+ * the stream before using the returned results. This pattern ensures safe memory lifetime
+ * management when chaining operations.
+ *
+ * @note Uses exec_policy_nosync internally - caller must synchronize before using results
+ */
+template <bool has_nulls>
+std::pair<rmm::device_uvector<cudf::size_type>, rmm::device_uvector<cudf::size_type>>
+filter_by_conditional_impl(rmm::device_uvector<cudf::size_type>& left_indices,
+                           rmm::device_uvector<cudf::size_type>& right_indices,
+                           cudf::table_device_view const& left_table,
+                           cudf::table_device_view const& right_table,
+                           cudf::ast::detail::expression_device_view device_expression_data,
+                           cudf::size_type num_intermediates,
+                           rmm::cuda_stream_view stream,
+                           rmm::device_async_resource_ref mr)
+{
+  auto const num_pairs = left_indices.size();
+  if (num_pairs == 0) {
+    return {rmm::device_uvector<cudf::size_type>(0, stream, mr),
+            rmm::device_uvector<cudf::size_type>(0, stream, mr)};
+  }
+
+  // Create a boolean mask for indices to keep
+  auto keep_mask = rmm::device_uvector<bool>(num_pairs, stream, mr);
+
+  // Launch kernel to evaluate expression for each pair with dynamic shared memory sizing
+  int current_device = 0;
+  CUDF_CUDA_TRY(cudaGetDevice(&current_device));
+  int max_shmem_per_block = 0;
+  CUDF_CUDA_TRY(cudaDeviceGetAttribute(
+    &max_shmem_per_block, cudaDevAttrMaxSharedMemoryPerBlock, current_device));
+
+  // Use the same type alias as in the kernel to ensure consistency
+  auto const per_thread_bytes = static_cast<int>(num_intermediates) *
+                                static_cast<int>(sizeof(intermediate_storage_type<has_nulls>));
+  int block_size = 256;  // default
+  if (per_thread_bytes > 0) {
+    int const max_by_shmem = max_shmem_per_block / per_thread_bytes;
+    if (max_by_shmem > 0) {
+      // Prefer multiples of warp size (32)
+      int const rounded = (max_by_shmem / 32) * 32;
+      block_size        = std::max(32, std::min(256, rounded > 0 ? rounded : max_by_shmem));
+    } else {
+      block_size = 32;  // minimal reasonable size
+    }
+  }
+  cudf::size_type const grid_size = (num_pairs + block_size - 1) / block_size;
```

Comment:
What if `num_pairs` is huge? This will result in a huge grid size. You will not have enough shared memory for all blocks, and the program will crash.

The `grid_size` should be evaluated based on both the amount of shared memory and input size. Please see how we do that here:
 * https://github.com/rapidsai/cudf/blob/b321bb117fdef4c2b17aef3874b68230702edee7/cpp/src/groupby/hash/compute_single_pass_aggs.cuh#L70-L79
 * https://github.com/rapidsai/cudf/blob/b321bb117fdef4c2b17aef3874b68230702edee7/cpp/src/groupby/hash/compute_shared_memory_aggs.cu#L287-L293